pkg/backup/manager.go

```diff
@@ -0,0 +1,313 @@
+	}
+
+	opts.DefaultImageNameTemplate, err = template.New("default-image-name").Parse(config.DefaultImageNameFormat())
+	if err != nil {
+		return opts, err
+	}
+
+	opts.DefaultMaxKeepImages, err = config.DefaultMaxKeepImages()
+	return opts, err
+}
+
+// Manager manages backups/images of ec2 resources(volumes, instances, etc.)
+type Manager struct {
+	client *ec2.EC2
+
+	volumes   []*ec2.Volume
+	instances []*ec2.Instance
+
+	BackupTagKey     string
+	BackupTagValue   string
+	ImageTagKey      string
+	ImageTagValue    string
+	ImageNameTag     string
+	MaxKeepImagesTag string
+
+	DefaultImageNameTemplate *template.Template
+	DefaultMaxKeepImages     int
+
+	Verbose bool
+}
+
+// NewManager creates a new backup manager from the provided options
+func NewManager(opts *ManagerOpts) (*Manager, error) {
+	m := &Manager{
+		volumes:   make([]*ec2.Volume, 0),
+		instances: make([]*ec2.Instance, 0),
+	}
+	m.client = opts.Client
+	m.BackupTagKey = opts.BackupTagKey
+	m.BackupTagValue = opts.BackupTagValue
+	m.ImageTagKey = opts.ImageTagKey
+	m.ImageTagValue = opts.ImageTagValue
+	m.ImageNameTag = opts.ImageNameTag
+	m.Verbose = opts.Verbose
+	if opts.DefaultImageNameTemplate == nil {
+		return nil, fmt.Errorf("DefaultImageNameTemplate is a required field for ManagerOpts")
+	}
+
+	m.DefaultImageNameTemplate = opts.DefaultImageNameTemplate
+	m.DefaultMaxKeepImages = opts.DefaultMaxKeepImages
+	return m, nil
+}
+
+// Search searches for a volumes and instances to backup
+func (m *Manager) Search() error {
+	return m.all(
+		[]func() error{
+			m.searchVolumes,
+			m.searchInstances,
+		},
+	)
+}
+
+func (m *Manager) searchVolumes() error {
+	params := &ec2.DescribeVolumesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String(fmt.Sprintf("tag:%s", m.BackupTagKey)),
+				Values: []*string{aws.String(m.BackupTagValue)},
+			},
+		},
+		MaxResults: aws.Int64(500),
+	}
+
+	m.logf("Searching for volumes with tag %s:%s\n", m.BackupTagKey, m.BackupTagValue)
+
+	return m.client.DescribeVolumesPages(params,
+		func(page *ec2.DescribeVolumesOutput, lastPage bool) bool {
+			for _, v := range page.Volumes {
+				m.volumes = append(m.volumes, v)
+			}
+			return !lastPage
+		})
+}
+
+func (m *Manager) searchInstances() error {
+	params := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String(fmt.Sprintf("tag:%s", m.ImageTagKey)),
+				Values: []*string{aws.String(m.ImageTagValue)},
+			},
+		},
+		MaxResults: aws.Int64(500),
+	}
+
+	m.logf("Searching for instances with tag %s:%s\n", m.ImageTagKey, m.ImageTagValue)
+
+	return m.client.DescribeInstancesPages(params,
+		func(page *ec2.DescribeInstancesOutput, lastPage bool) bool {
+			for _, r := range page.Reservations {
+				for _, i := range r.Instances {
+					tags := utils.TagSliceToMap(i.Tags)
+					m.logf(
+						"Found instance %s(%s) with matching tag\n",
+						aws.StringValue(i.InstanceId),
+						tags.GetDefault("Name", "<no value>"),
```

Comment:
🌵 Could this default to `<no 'Name' value>`?

Reply:
I'd prefer to stick with `<no value>`. Seeing as how it you can go back to the logs to see. In most of the dealings with these, I've most commonly seen `<nil>` or `<no value>`.